src/atlas/repartition/detail/StructuredColumnsToStructuredColumns.h

```diff
@@ -0,0 +1,144 @@
+/*
+ * (C) Crown Copyright 2021 Met Office
+ *
+ * This software is licensed under the terms of the Apache Licence Version 2.0
+ * which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+ */
+
+#pragma once
+
+#include <vector>
+
+#include "atlas/functionspace/StructuredColumns.h"
+#include "atlas/repartition/detail/RepartitionImpl.h"
+
+namespace atlas {
+
+  class Field;
+  class FieldSet;
+  class FunctionSpace;
+
+  namespace functionspace {
+    namespace detail {
+      class StructuredColumns;
+    }
+  }
+}
+
+namespace atlas {
+  namespace repartition {
+    namespace detail {
+
+      // Forward declarations.
+      class StructuredColumnsToStructuredColumns;
+      class FuncSpaceRange;
```

Comment:
Please rename `FuncSpaceRange` to `detail::StructuredIndexRange` to be more unique and descriptive to what it applies

Reply:
Good point. I originally called it `IndexRange`, then I learned that this was a thing elsewhere. In my panic, I renamed it something rubbish! This is a better name.

Reply:
> Hi @odlomax , Very high quality code!
> Please address a few minor cosmetic changes about naming.
> 
> Did you check if redistribution works with different values for halo for source and target?

Thanks @wdeconinck  for your kind words! I've addressed your comments and added a couple tests with mixed halo sizes.